### PR TITLE
[#195] Fix parser exponential behavior

### PR DESCRIPTION
**Pull request by jiegillet**
Fixed the exponential behavior, I've had to change a bunch of functions, but in the end it let me refactor `tomlP` nicely so I think it's worth it. Benchmarking also improved.

Since some signatures changed, I've had to modify some tests.

I've also added an `eof` to the TOML parser to make sure the whole `.toml` file is valid.

Benchmarks:
Without fix:
```
benchmarked Parse/tomland
time                 440.8 μs   (430.0 μs .. 453.9 μs)
                     0.993 R²   (0.989 R² .. 0.997 R²)
mean                 433.0 μs   (427.9 μs .. 440.4 μs)
std dev              20.27 μs   (14.42 μs .. 31.25 μs)
variance introduced by outliers: 27% (moderately inflated)
```

With fix:
```
benchmarked Parse/tomland
time                 355.8 μs   (348.9 μs .. 368.5 μs)
                     0.990 R²   (0.980 R² .. 0.998 R²)
mean                 360.6 μs   (356.1 μs .. 368.5 μs)
std dev              19.44 μs   (13.67 μs .. 27.89 μs)
variance introduced by outliers: 32% (moderately inflated)
```
That's a 15% speedup.

### CODE REVIEW

**Comment by jiegillet** [comment]
Some checks are failing: `htoml-megaparsec` doesn't build on `8.4.4`, not sure what I can do about that.

File: test/Test/Toml/Parsing/Unit.hs [hunk 1]
```diff
@@ -396,49 +396,39 @@ tableSpecs :: Spec
...
-                (makeKey ["table-1"], tomlFromKeyVal [str, int])
+          parseTable "key1 = \"some string\"\nkey2 = 123"
+                (tomlFromKeyVal [str, int])
         it "can parse an empty TOML table" $
-            parseTable "[table]" (makeKey ["table"], tomlFromKeyVal [])
-        it "allows the name of the table to be any valid TOML key" $ do
-            parseTable "[dog.\"tater.man\"]"
-                (makeKey ["dog", dquote "tater.man"], tomlFromKeyVal [])
-            parseTable "[j.\"ʞ\".'l']"
-                (makeKey ["j", dquote "ʞ", squote "l"], tomlFromKeyVal [])
```
Comment (chshersh) [inline test/Test/Toml/Parsing/Unit.hs]: I don't quite understand why so many tests were deleted :thinking: They seem useful because they test unicode symbols. It's not obvious from the implementation, why these tests are redundant now.

File: src/Toml/Parser/TOML.hs [hunk 2]
```diff
@@ -68,69 +68,58 @@ inlineTableP = between
...
     (table, array) <- fmap distributeEithers
         $ many
-        $ childKeyP key
-        $ eitherPairP (try tableP) tableArrayP
+        $ eitherPairP
+            (try $ do
+              (kDiff, k) <- childKeyP key tableNameP
+              t <- tableP (Just k)
+              pure (kDiff, t))
+            (do
```
Comment (chshersh) [inline src/Toml/Parser/TOML.hs]: I think it's better for code readability to introduce two local variables under `let` with types and names like `tomlTableP` and `tomlTableArrayP` and use them as arguments for `eitherPairP` instead of manually wrapping `do` blocks in `()`.

File: src/Toml/Parser/TOML.hs [hunk 3]
```diff
@@ -68,69 +68,58 @@ inlineTableP = between
...
+tableP :: Maybe Key -> Parser TOML
+tableP key = do
     (val, inline)  <- distributeEithers <$> many hasKeyP
     (table, array) <- fmap distributeEithers
         $ many
-        $ childKeyP key
-        $ eitherPairP (try tableP) tableArrayP
+        $ eitherPairP
+            (try $ do
+              (kDiff, k) <- childKeyP key tableNameP
```
Comment (chshersh) [inline src/Toml/Parser/TOML.hs]: It looks like this parser can hit another corner case because it has backtracking due to nested `try` combinators inside self-recursive calls. My understanding is that you should put `try` as deep as possible. I see that `childKeyP` already has `try`. Would the tests still pass if you just remove that `try` from the line above (88)?

File: src/Toml/Parser/TOML.hs [hunk 4]
```diff
@@ -68,69 +68,58 @@ inlineTableP = between
...
-childKeyP :: Key -> Parser (Key, a) -> Parser (Key, a)
-childKeyP key p = try $ do
-    (k, x) <- p
+-- | @childKeyP (Just key) p@ checks if the result of @p@ if a child key of
+-- @key@ and returns the difference of the keys and the child key.
+-- @childKeyP Nothing p@ is only called from @tomlP@ (no parent key).
+childKeyP :: Maybe Key -> Parser Key -> Parser (Key, Key)
+childKeyP Nothing parser = try $ do
+  k <- parser
+  pure (k, k)
```
Comment (chshersh) [inline src/Toml/Parser/TOML.hs]: I would rewrite it as `try $ (\x -> (x, x) <$> parser`. I see why you did it this way: you prefer symmetry in pattern-matching. In that case, you can leave the implementation as it is, but you should use 4 spaces according to our style guide.

File: src/Toml/Parser/TOML.hs (hunk 2 above)
Comment (jiegillet) [reply src/Toml/Parser/TOML.hs]: You're right, I'll change it

File: src/Toml/Parser/TOML.hs (hunk 3 above)
Comment (jiegillet) [reply src/Toml/Parser/TOML.hs]: I'll check!

File: src/Toml/Parser/TOML.hs (hunk 4 above)
Comment (jiegillet) [reply src/Toml/Parser/TOML.hs]: I don't mind your suggestion, I didn't particularly like my way of writing it. And I'll re-check the spacing everywhere.

**Comment by chshersh** [comment]
@jiegillet Ci should pass now. I cleaned the cache and restarted it. Not sure what's happening here exactly and why this error happens.

File: test/Test/Toml/Parsing/Unit.hs (hunk 1 above)
Comment (jiegillet) [reply test/Test/Toml/Parsing/Unit.hs]: I had to remove them because I've refactored the key parsing out of the table parsing function. Those exact same tests on unicode keys are tested somewhere else, so they were redundant anyway.
I will add more tests like you suggested though. There are also some corner cases that I should add.

**Comment by jiegillet** [comment]
Sorry I ended up moving a lot of things around in tests. I ditched all the partial tables and arrays tests in favor of full on `tomlP` tests. I didn't remove any test, and I added one with many arrays.
